### ISSUE TITLE
BatchNorm in OP_SET version 7 has no mode attribute

### ISSUE THREAD

**Issue by anirudhacharya** [onnx/onnx#1042]
The [BatchNormalization](https://github.com/onnx/onnx/blob/master/docs/Operators.md#batchnormalization) operator does not have ``mode`` attribute, so how do we figure out the difference between 'training' and 'testing' mode.

The [OP_SET version 6](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#batchnormalization-6) of the operator had a ``is_test`` attribute.

@bddppq @spidyDev

**Comment by prasanthpul** [comment]
This seems to have been done in PR #1014. @linkerzhang please take a look

**Comment by linkerzhang** [comment]
Yes, "is_test" is not needed as we're following same "optional" input/output semantics across whole op spec. PR #1014 and PR #1006 are both for this unifying purpose. The "optional" semantics is documented as https://github.com/onnx/onnx/blob/master/docs/IR.md#optional-inputs-and-outputs.

**Comment by tjingrant** [comment]
@linkerzhang @prasanthpul @anirudhacharya @bddppq  I came across the same issue. I don't think "is_test" is solely used to indicate the number of outputs; it also mandates how we treat running_mean and running_var. Namely, statistics of the mini-batch in inference mode is not updated into the running mean/var.

I quote the batch norm paper https://arxiv.org/abs/1502.03167, section 3.1:
> Since the means and variances are fixed during inference, the normalization is simply a linear transform applied to each activation. 

Am I missing anything?

**Comment by ebarsoum** [comment]
Yes, testing vs training mode is not just about the number of outputs, it define the behavior of the node (BN, dropout and others). However, currently we only support inference so by definition is_test is always true. We are missing this attributes in other OPs also. The training track should look at all existing OPs and properly add training or testing flag. Default value or non existing should mean inference.

**Comment by tjingrant** [comment]
@ebarsoum keeping two output cases (train_mode/test_mode) and removing the train/test flag does not sound very logically sanitary. Can we make a decision and either:
1. Have two output cases and the train/test flag.
2. Or remove optional outputs along with the is_test flag.

I'm okay with both, but being stuck in the middle could be a source of confusion and trouble.

**Comment by tjingrant** [comment]
@ebarsoum it seems like the assumption that we are always in inference mode goes directly contradictory to a default momentum value of 0.9. In inference mode, momentum is effectively 1.0. How should we reconcile these contradictory default values?

**Comment by houseroad** [comment]
@tjingrant in the test mode, the batchnorm does not really use `momentum`, because the weights are not updated. I agree with that if we only support inference mode, the output number should be 1 and some attributes are not necessary.

**Comment by houseroad** [comment]
For now, you can safely assume that we only support inference. The plan of training support is unclear at this moment. With current spec, I think you should be able to express your model in eval mode. (Although there are unnecessary optional attributes and output)

**Comment by pranavsharma** [comment]
I came across the removal of is_test recently. The optional input/output semantics has been cited as a reason for removing this flag. I don't think these semantics can be applied as-is in this case because it is not just about the number of outputs that the BatchNorm kernel creates; its about the behavior of the kernel that changes in training mode vs inference mode. The optional input/output semantics argument would apply if we're not going to support training at all. Moreover if we're not going to keep is_test, we should remove references to momentum, spatial and multiple outputs as well since they apply to training only and it only adds confusion to the spec.

**Comment by SherlockNoMad** [comment]
I have open a PR to make Dropout and BatchNormalization Training-friendly. This PR should also address this issue.

**Comment by snnn** [comment]
The latest spec still confusing to me. 
According to the test cases, momentum is ignored in inference mode but it is default to 0.9 in training mode. However, neither the node nor the model has a training/inference mode setting. They just left the choice to frameworks.  So the issue is not solved yet.

**Comment by pranav-prakash** [comment]
@snnn The `training_mode` flag was reverted as part of https://github.com/onnx/onnx/pull/2750
So as of the moment the situation doesn't seem to be any better than it was at the start of the discussion thread; one has to infer training mode based on the existence of optional outputs, but this is not bulletproof and can possibly lead to unexpected behavior.

**Comment by gramalingam** [comment]
@pranav-prakash : a question: is there a use-case where we want to use optional outputs in the inference-mode? Just trying to understand what the desired semantics is in the inference-mode, and how it is required to be different from the training-mode (other than the optional outputs being computed)?

**Comment by pranav-prakash** [comment]
@gramalingam There shouldn't ever be a need to use those optional outputs in inference mode. However, such a model is currently permitted per the ONNX spec, so even though I agree that it is an unlikely pathological edge-case that shouldn't happen in well-behaved models, the spec should either be updated to include an explicit attribute for training mode or state that if optional outputs are present the behavior will be training mode.

**Comment by gramalingam** [comment]
Thanks. I am trying to understand what we need: (a) A clarification that the presence of optional outputs means training mode, or (b) An extra attribute to specify the training mode. When I look at the training_mode flag in the reverted PR (linked in your message above), I cannot immediately figure out the need for the flag. Wondering if that flag was just to be extra-safe and avoid ambiguity.

**Comment by pranav-prakash** [comment]
In fact even in the existing onnx spec it states "The running mean after the BatchNormalization operator." in the description of the optional mean output – those words "running mean" only make sense if it's in training mode. So going with option (a) and adding clarification should not change anything.

**Comment by askhade** [comment]
Closing this issue since BatchNorm has been updated now to include the training mode : https://github.com/onnx/onnx/blob/master/docs/Operators.md#BatchNormalization
